Spurrius error while tracking downloadable files

Here is a recent test failure:

```

___________________________ test_filter_track_multi ____________________________

datafiles = PosixPath('/home/testuser/buildstream/.tox/py311/tmp/test_filter_track_multi0')
cli = <buildstream._testing.runcli.Cli object at 0x7fb6d291fb90>
tmpdir = local('/home/testuser/buildstream/.tox/py311/tmp/test_filter_track_multi0')

    @pytest.mark.datafiles(os.path.join(DATA_DIR, "basic"))
    def test_filter_track_multi(datafiles, cli, tmpdir):
        repo = create_repo("tar", str(tmpdir))
        ref = repo.create(os.path.join(str(datafiles), "files"))
        elements_dir = os.path.join(str(tmpdir), "elements")
        project = str(tmpdir)
        input_name = "input.bst"
        input2_name = "input2.bst"
    
        project_config = {
            "name": "filter-track-test",
            "min-version": "2.0",
            "element-path": "elements",
        }
        project_file = os.path.join(str(tmpdir), "project.conf")
        _yaml.roundtrip_dump(project_config, project_file)
    
        input_config = {
            "kind": "import",
            "sources": [repo.source_config()],
        }
    
        input_file = os.path.join(elements_dir, input_name)
        _yaml.roundtrip_dump(input_config, input_file)
    
        input2_config = dict(input_config)
        input2_file = os.path.join(elements_dir, input2_name)
        _yaml.roundtrip_dump(input2_config, input2_file)
    
        filter1_config = {"kind": "filter", "depends": [{"filename": input_name, "type": "build"}]}
        filter1_file = os.path.join(elements_dir, "filter1.bst")
        _yaml.roundtrip_dump(filter1_config, filter1_file)
    
        filter2_config = {"kind": "filter", "depends": [{"filename": input2_name, "type": "build"}]}
        filter2_file = os.path.join(elements_dir, "filter2.bst")
        _yaml.roundtrip_dump(filter2_config, filter2_file)
    
        # Assert that a fetch is needed
        states = cli.get_element_states(project, [input_name, input2_name])
    
        assert states == {
            input_name: "no reference",
            input2_name: "no reference",
        }
    
        # Now try to track it
        result = cli.run(project=project, args=["source", "track", "filter1.bst", "filter2.bst"])
>       result.assert_success()

tests/elements/filter.py:391: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <buildstream._testing.runcli.Result object at 0x7fb6d28e3750>
fail_message = ''

    def assert_success(self, fail_message=""):
>       assert self.exit_code == 0, fail_message
E       AssertionError

src/buildstream/_testing/runcli.py:104: AssertionError
----------------------------- Captured stdout call -----------------------------
BuildStream exited with code 0 for invocation:
	bst --no-colors --no-verbose --config /home/testuser/buildstream/.tox/py311/tmp/test_filter_track_multi0/cache/buildstream.conf --directory /home/testuser/buildstream/.tox/py311/tmp/test_filter_track_multi0 show --deps all --format %{name}||%{state} input.bst input2.bst
Program output was:
input.bst||no reference
input2.bst||no reference

Program stderr was:
[--:--:--][        ][    main:core activity                 ] START   Loading elements
[00:00:00][        ][    main:core activity                 ] SUCCESS Loading elements

    2 subtasks processed

[--:--:--][        ][    main:core activity                 ] START   Resolving elements
[--:--:--][        ][    main:input.bst                     ] WARNING [unaliased-url]: tar source at input.bst [line 3 column 2]: Use of unaliased source download URL: file:///home/testuser/buildstream/.tox/py311/tmp/test_filter_track_multi0/repo/file.tar.gz
[--:--:--][        ][    main:input2.bst                    ] WARNING [unaliased-url]: tar source at input2.bst [line 3 column 2]: Use of unaliased source download URL: file:///home/testuser/buildstream/.tox/py311/tmp/test_filter_track_multi0/repo/file.tar.gz
[00:00:00][        ][    main:core activity                 ] SUCCESS Resolving elements

    2 of 2 subtasks processed

[--:--:--][        ][    main:core activity                 ] START   Initializing remote caches
[00:00:00][        ][    main:core activity                 ] SUCCESS Initializing remote caches
[--:--:--][        ][    main:core activity                 ] START   Query cache
[00:00:00][        ][    main:core activity                 ] SUCCESS Query cache

    2 of 2 subtasks processed


BuildStream exited with code -1 for invocation:
	bst --no-colors --config /home/testuser/buildstream/.tox/py311/tmp/test_filter_track_multi0/cache/buildstream.conf --directory /home/testuser/buildstream/.tox/py311/tmp/test_filter_track_multi0 source track filter1.bst filter2.bst
Program stderr was:
[--:--:--][        ][    main:core activity                 ] START   Track
[--:--:--][        ][    main:core activity                 ] START   Loading elements
[00:00:00][        ][    main:core activity                 ] SUCCESS Loading elements

    4 subtasks processed

[--:--:--][        ][    main:core activity                 ] START   Resolving elements
[--:--:--][        ][    main:input.bst                     ] WARNING [unaliased-url]: tar source at input.bst [line 3 column 2]: Use of unaliased source download URL: file:///home/testuser/buildstream/.tox/py311/tmp/test_filter_track_multi0/repo/file.tar.gz
[--:--:--][        ][    main:input2.bst                    ] WARNING [unaliased-url]: tar source at input2.bst [line 3 column 2]: Use of unaliased source download URL: file:///home/testuser/buildstream/.tox/py311/tmp/test_filter_track_multi0/repo/file.tar.gz
[00:00:00][        ][    main:core activity                 ] SUCCESS Resolving elements

    4 of 4 subtasks processed


BuildStream Version 2.0.1+7.g5be4ed70f
    Session Start: Saturday, 04-03-2023 at 05:04:22
    Project:       filter-track-test (/home/testuser/buildstream/.tox/py311/tmp/test_filter_track_multi0)
    Targets:       filter1.bst, filter2.bst

User Configuration
    Configuration File:      /home/testuser/buildstream/.tox/py311/tmp/test_filter_track_multi0/cache/buildstream.conf
    Cache Directory:         /home/testuser/buildstream/.tox/py311/tmp/test_filter_track_multi0/cache
    Log Files:               /home/testuser/buildstream/.tox/py311/tmp/test_filter_track_multi0/cache/logs
    Source Mirrors:          /home/testuser/buildstream/.tox/py311/tmp/test_filter_track_multi0/cache/sources
    Build Area:              /home/testuser/buildstream/.tox/py311/tmp/test_filter_track_multi0/cache/build
    Strict Build Plan:       Yes
    Maximum Fetch Tasks:     10
    Maximum Build Tasks:     4
    Maximum Push Tasks:      4
    Maximum Network Retries: 2

Project: filter-track-test

    Element Plugins
        filter: core plugin
        import: core plugin

    Source Plugins
        tar: core plugin

Pipeline
no reference ???????????????????????????????????????????????????????????????? input.bst 
     waiting ???????????????????????????????????????????????????????????????? filter1.bst 
no reference ???????????????????????????????????????????????????????????????? input2.bst 
     waiting ???????????????????????????????????????????????????????????????? filter2.bst 
===============================================================================
[--:--:--][????????][   track:input.bst                     ] START   filter-track-test/input/????????-track.251.log
[--:--:--][????????][   track:input.bst                     ] START   Tracking file:///home/testuser/buildstream/.tox/py311/tmp/test_filter_track_multi0/repo/file.tar.gz
[--:--:--][????????][   track:input2.bst                    ] START   filter-track-test/input2/????????-track.251.log
[--:--:--][????????][   track:input2.bst                    ] START   Tracking file:///home/testuser/buildstream/.tox/py311/tmp/test_filter_track_multi0/repo/file.tar.gz
[00:00:00][????????][   track:input2.bst                    ] SUCCESS Tracking file:///home/testuser/buildstream/.tox/py311/tmp/test_filter_track_multi0/repo/file.tar.gz
[00:00:00][????????][   track:input.bst                     ] SUCCESS Tracking file:///home/testuser/buildstream/.tox/py311/tmp/test_filter_track_multi0/repo/file.tar.gz
[--:--:--][????????][   track:input2.bst                    ] INFO    Found new revision: ef9de6877541ca4570fb67463a34678c443cbfac82b4c16740d08215dd6194e7
[00:00:00][????????][   track:input2.bst                    ] SUCCESS filter-track-test/input2/????????-track.251.log
[00:00:00][????????][   track:input.bst                     ] BUG     Track

    An unhandled exception occured:
    
    Traceback (most recent call last):
      File "/home/testuser/buildstream/src/buildstream/_scheduler/jobs/job.py", line 438, in child_action
        result = self.child_process()  # pylint: disable=assignment-from-no-return
                 ^^^^^^^^^^^^^^^^^^^^
      File "/home/testuser/buildstream/src/buildstream/_scheduler/jobs/elementjob.py", line 91, in child_process
        return self._action_cb(self._element)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/testuser/buildstream/src/buildstream/_scheduler/queues/trackqueue.py", line 65, in _track_element
        return element._track()
               ^^^^^^^^^^^^^^^^
      File "/home/testuser/buildstream/src/buildstream/element.py", line 1390, in _track
        return self.__sources.track(self._get_workspace())
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/testuser/buildstream/src/buildstream/_elementsources.py", line 97, in track
        new_ref = source._track()
                  ^^^^^^^^^^^^^^^
      File "/home/testuser/buildstream/src/buildstream/source.py", line 1202, in _track
        new_ref = self.__do_track()
                  ^^^^^^^^^^^^^^^^^
      File "/home/testuser/buildstream/src/buildstream/source.py", line 1430, in __do_track
        return self.track(**kwargs)
               ^^^^^^^^^^^^^^^^^^^^
      File "/home/testuser/buildstream/src/buildstream/downloadablefilesource.py", line 168, in track
        new_ref = self._ensure_mirror("Tracking {}".format(self.url))
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/testuser/buildstream/src/buildstream/downloadablefilesource.py", line 239, in _ensure_mirror
        os.makedirs(self._mirror_dir)
      File "<frozen os>", line 225, in makedirs
    FileExistsError: [Errno 17] File exists: '/home/testuser/buildstream/.tox/py311/tmp/test_filter_track_multi0/cache/sources/tar/file____home_testuser_buildstream__tox_py311_tmp_test_filter_track_multi0_repo_file_tar_gz'

[00:00:00][        ][    main:core activity                 ] FAILURE Track

Resolved key Summary
     waiting 4b64454b13380afb2a6fd2ced56cd1de8b6a7454f4972c81cbac01aed70210be input2.bst 

Failure Summary
    input.bst:
    [00:00:00][????????][   track:input.bst                     ] BUG     Track

        An unhandled exception occured:
    
        Traceback (most recent call last):
          File "/home/testuser/buildstream/src/buildstream/_scheduler/jobs/job.py", line 438, in child_action
            result = self.child_process()  # pylint: disable=assignment-from-no-return
                     ^^^^^^^^^^^^^^^^^^^^
          File "/home/testuser/buildstream/src/buildstream/_scheduler/jobs/elementjob.py", line 91, in child_process
            return self._action_cb(self._element)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
          File "/home/testuser/buildstream/src/buildstream/_scheduler/queues/trackqueue.py", line 65, in _track_element
            return element._track()
                   ^^^^^^^^^^^^^^^^
          File "/home/testuser/buildstream/src/buildstream/element.py", line 1390, in _track
            return self.__sources.track(self._get_workspace())
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
          File "/home/testuser/buildstream/src/buildstream/_elementsources.py", line 97, in track
            new_ref = source._track()
                      ^^^^^^^^^^^^^^^
          File "/home/testuser/buildstream/src/buildstream/source.py", line 1202, in _track
            new_ref = self.__do_track()
                      ^^^^^^^^^^^^^^^^^
          File "/home/testuser/buildstream/src/buildstream/source.py", line 1430, in __do_track
            return self.track(**kwargs)
                   ^^^^^^^^^^^^^^^^^^^^
          File "/home/testuser/buildstream/src/buildstream/downloadablefilesource.py", line 168, in track
            new_ref = self._ensure_mirror("Tracking {}".format(self.url))
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
          File "/home/testuser/buildstream/src/buildstream/downloadablefilesource.py", line 239, in _ensure_mirror
            os.makedirs(self._mirror_dir)
          File "<frozen os>", line 225, in makedirs
        FileExistsError: [Errno 17] File exists: '/home/testuser/buildstream/.tox/py311/tmp/test_filter_track_multi0/cache/sources/tar/file____home_testuser_buildstream__tox_py311_tmp_test_filter_track_multi0_repo_file_tar_gz'

Pipeline Summary
    Total:       4
    Session:     2
    Track Queue: processed 1, skipped 0, failed 1 
```

--- a/src/buildstream/downloadablefilesource.py
+++ b/src/buildstream/downloadablefilesource.py
@@ -235,8 +235,12 @@ class DownloadableFileSource(Source):
                 return self.ref
 
             # Make sure url-specific mirror dir exists.
-            if not os.path.isdir(self._mirror_dir):
-                os.makedirs(self._mirror_dir)
+            try:
+                os.makedirs(self._mirror_dir, exist_ok=True)
+            except FileExistsError as e:
+                raise SourceError(
+                    "{}: Mirror directory exists but is not a directory: {}".format(self, self._mirror_dir)
+                ) from e
 
             # Store by sha256sum
             sha256 = utils.sha256sum(local_file)

--- a/tests/elements/filter.py
+++ b/tests/elements/filter.py
@@ -347,8 +347,6 @@ def test_filter_track_multi(datafiles, cli, tmpdir):
     ref = repo.create(os.path.join(str(datafiles), "files"))
     elements_dir = os.path.join(str(tmpdir), "elements")
     project = str(tmpdir)
-    input_name = "input.bst"
-    input2_name = "input2.bst"
 
     project_config = {
         "name": "filter-track-test",
@@ -363,43 +361,41 @@ def test_filter_track_multi(datafiles, cli, tmpdir):
         "sources": [repo.source_config()],
     }
 
-    input_file = os.path.join(elements_dir, input_name)
-    _yaml.roundtrip_dump(input_config, input_file)
+    #
+    # Lets do 10 of them at the same time, this may help regression
+    # test against https://github.com/apache/buildstream/issues/1831
+    #
+    filter_elements = []
+    input_elements = []
+    for num in range(0, 10):
+        input_elt = f"input{num}.bst"
+        input_file = os.path.join(elements_dir, input_elt)
+        _yaml.roundtrip_dump(input_config, input_file)
 
-    input2_config = dict(input_config)
-    input2_file = os.path.join(elements_dir, input2_name)
-    _yaml.roundtrip_dump(input2_config, input2_file)
+        filter_config = {"kind": "filter", "depends": [{"filename": input_elt, "type": "build"}]}
+        filter_elt = f"filter{num}.bst"
+        filter_file = os.path.join(elements_dir, filter_elt)
+        _yaml.roundtrip_dump(filter_config, filter_file)
 
-    filter1_config = {"kind": "filter", "depends": [{"filename": input_name, "type": "build"}]}
-    filter1_file = os.path.join(elements_dir, "filter1.bst")
-    _yaml.roundtrip_dump(filter1_config, filter1_file)
-
-    filter2_config = {"kind": "filter", "depends": [{"filename": input2_name, "type": "build"}]}
-    filter2_file = os.path.join(elements_dir, "filter2.bst")
-    _yaml.roundtrip_dump(filter2_config, filter2_file)
+        input_elements.append(input_elt)
+        filter_elements.append(filter_elt)
 
     # Assert that a fetch is needed
-    states = cli.get_element_states(project, [input_name, input2_name])
-
-    assert states == {
-        input_name: "no reference",
-        input2_name: "no reference",
-    }
+    states = cli.get_element_states(project, input_elements)
+    for key in states:
+        assert states[key] == "no reference"
 
     # Now try to track it
-    result = cli.run(project=project, args=["source", "track", "filter1.bst", "filter2.bst"])
+    result = cli.run(project=project, args=["source", "track"] + filter_elements)
     result.assert_success()
 
     # Now check that a ref field exists
-    new_input = _yaml.load(input_file, shortname=None)
-    source_node = new_input.get_sequence("sources").mapping_at(0)
-    new_ref = source_node.get_str("ref")
-    assert new_ref == ref
-
-    new_input2 = _yaml.load(input2_file, shortname=None)
-    source_node2 = new_input2.get_sequence("sources").mapping_at(0)
-    new_ref2 = source_node2.get_str("ref")
-    assert new_ref2 == ref
+    for elt in input_elements:
+        input_file = os.path.join(elements_dir, elt)
+        new_input = _yaml.load(input_file, shortname=None)
+        source_node = new_input.get_sequence("sources").mapping_at(0)
+        new_ref = source_node.get_str("ref")
+        assert new_ref == ref
 
 
 @pytest.mark.datafiles(os.path.join(DATA_DIR, "basic"))

